apps/chat/agent/tools.py

```diff
@@ -124,24 +124,21 @@ def _move_datetime_to_new_weekday_and_time(date: datetime, new_weekday: int, new
     day_diff = new_weekday - current_weekday
     return date.replace(hour=new_hour, minute=new_minute, second=0) + timedelta(days=day_diff)
 
-
-def create_periodic_task(experiment_session: ExperimentSession, message: str, **kwargs):
-    task_kwargs = json.dumps(
-        {
-            "chat_ids": [experiment_session.participant.identifier],
-            "message": message,
-            "is_bot_instruction": False,
-            "experiment_public_id": str(experiment_session.experiment.public_id),
-        }
-    )
-    PeriodicTask.objects.create(
-        name=f"reminder-{experiment_session.id}-{uuid.uuid4()}",
-        task=BOT_MESSAGE_FOR_USER_TASK,
-        kwargs=task_kwargs,
-        **kwargs,
+def create_schedule_message(experiment_session: ExperimentSession, message: str, trigger_date: datetime, **kwargs):
+    from apps.utils.factories.events import EventActionFactory
+    additional_params = {
+        "name": f"reminder-{experiment_session.id}-{uuid.uuid4()}",
+        'prompt_text': message,
+    }
+    params = {**kwargs, **additional_params}
```

Comment:
let's rather be explicit in which keys we put in the params for the Event, since we expect those that are in [this form](https://github.com/dimagi/open-chat-studio/blob/bfc1c1b0c5732286400789c8f447526c283f6a15/apps/events/forms.py#L65) to be in there. I quite liked what you did previously in the migration where you used the `ScheduledMessageConfigForm` to check the params.

What if, we treat the bot as just another type of user that needs to fill out the the `ScheduledMessageConfigForm` form? :thinking: So when we have all the necessary params, we then check the form's validity and if it's invalid, we tell the bot that it's not valid (we can do this by simply returning some string instead of the default "Success"). Hope this makes sense? Thoughts?

Reply:
A way to verify if we calculated the parameters for the form correctly is to do [this calculation](https://github.com/dimagi/open-chat-studio/blob/bfc1c1b0c5732286400789c8f447526c283f6a15/apps/events/models.py#L261) when we have the params. If the `next_trigger_date` matches the `datetime_due`, then we know we did it right

Reply:
Apologies, I'm using this as a brain dump :see_no_evil: 

Stepping back a bit makes me wonder: There are now two possible ways to create a scheduled message.
1. Experiment admins basically decide when/how it should be created. This is for a specific purpose i.e. the bot checking in on the user for some specific reason or to achive some goal.
2. The user asking the bot to check in on them or remind them about something. This is differnet though, because the user is the author and should have control over this schedule.

The the first case, it's an experiment admin that created the scheduled message indirectly through an event and the participant has some control over the schedule for the sake of preference i.e. they can ask for another day/time, but not move it out by say a week/month/year. Why? Because it's there for a reason.

Now in the second case, the participant is the one who authored the scheduled message, so it would make sense for them to be able to CRUD this schedule any way they see fit. When these changes are made, they will be able to create a schedule, but their ability to update/delete it will be limited. What if they asked the bot to create a reminder, but want to move it say a month later? We'll need to address this in a separate PR though. cc @snopoke

Reply:
Adding some context from the standup: I think we should combine the tools so that the same set are used for custom vs 'managed' schedules. In future we could provide options for experiment admins to restrict the changes allowed for managed schedules.

code notes:

`ScheduledMessage.action` can be made optional. Its purpose is to provide a reference to the params from the trigger.
For schedule messages that are created (or customised) by the user we can store params in `ScheduledMessage.custom_schedule_params`

When the ‘message’ is triggered we can use any values in `custom_schedule_params` to override the ones in the action (if there is an action)